daemon/hooks/hooks.go

```diff
@@ -0,0 +1,125 @@
+package hooks
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"time"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"bufio"
+	"bytes"
+
+	"github.com/zrepl/zrepl/logger"
+)
+
+type contextKey int
+
+const (
+	contextKeyLog contextKey = 0
+)
+
+type Logger = logger.Logger
+
+func WithLogger(ctx context.Context, log Logger) context.Context {
+	return context.WithValue(ctx, contextKeyLog, log)
+}
+
+func getLogger(ctx context.Context) Logger {
+	if log, ok := ctx.Value(contextKeyLog).(Logger); ok {
+		return log
+	}
+	return logger.NewNullLogger()
+}
+
+type logWriter struct {
+	mu *sync.Mutex
+	buf bytes.Buffer
+	scanner *bufio.Scanner
+	logFunc func(string)
+	level logger.Level
+}
+
+func NewLogWriter(mu *sync.Mutex, logFunc func(string)) *logWriter {
+	w := new(logWriter)
+	w.mu = mu
+	w.scanner = bufio.NewScanner(&w.buf)
+	w.logFunc = logFunc
+	return w
+}
+
+func (w *logWriter) Write(in []byte) (int, error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	bufCount, bufErr := w.buf.Write(in)
+	if bufErr != nil {
+		return bufCount, bufErr
+	}
+
+	if w.scanner.Scan() {
```

Comment:
should probably be a while loop, we are only going to log one line from the buffer otherwise, but the hook script could have logged several

Reply:
Hmm. Thinking about how to do this, because if `Write()` calls `Scan()` exhaustively, then the Scanner terminates prematurely and returns false forever. I'm thinking that I need to use something other than the Scanner here.